File: Python/0098-m-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


"""
Recursion - Divide and Conquer
Time: O(n)
Space: O(1)
"""
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        if not root:
            return True
        return self.isValidNode(root, float('-inf'), float('inf'))
    
    def isValidNode(self, cur, min, max):
        if not cur:
            return True
        if cur.val >= max or cur.val <= min:
            return False
        return self.isValidNode(cur.left, min, cur.val) and self.isValidNode(cur.right, cur.val, max)

File: Java/0098-m-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

/*
 * Recursion - Divide and Conquer
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        if (root == null)
            return true;
        return isValidNode(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }
    
    private boolean isValidNode(TreeNode root, long min, long max) {
        if (root == null)
            return true;
        if (root.val >= max || root.val <= min)
            return false;
        return isValidNode(root.left, min, root.val) && isValidNode(root.right, root.val, max);
    }
}